// frida/dist/device_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const device_1 = require("./device");
const signals_1 = require("./signals");
const util_1 = require("util");
class DeviceManager {
    constructor(impl) {
        this.impl = impl;
        const signals = new DeviceManagerSignals(impl.signals);
        this.added = new signals_1.Signal(signals, "added");
        this.removed = new signals_1.Signal(signals, "removed");
        this.changed = new signals_1.Signal(signals, "changed");
    }
    async enumerateDevices() {
        const devices = await this.impl.enumerateDevices();
        return devices.map(impl => new device_1.Device(impl));
    }
    async addRemoteDevice(host) {
        return new device_1.Device(await this.impl.addRemoteDevice(host));
    }
    removeRemoteDevice(host) {
        return this.impl.removeRemoteDevice(host);
    }
    [util_1.inspect.custom]() {
        return "DeviceManager {}";
    }
}
exports.DeviceManager = DeviceManager;
class DeviceManagerSignals extends signals_1.SignalAdapter {
    constructor(signals) {
        super(signals);
    }
    getProxy(name, userHandler) {
        if (name === "added" || name === "removed") {
            return impl => userHandler(new device_1.Device(impl));
        }
        return null;
    }
}
